File: api_gateway/internal/resolvers/viewer_protocol.go
package resolvers

import (
	"fmt"

	"github.com/Livepeer-FrameWorks/monorepo/pkg/mist"
	sharedpb "github.com/Livepeer-FrameWorks/monorepo/pkg/proto/shared"
)

func viewerProtocolRequirement(requested string) (string, error) {
	switch requested {
	case "":
		return "", nil
	case "WEBRTC", "WHEP", "HLS", "DASH", "HLS_CMAF", "MEWS", "MEWS_WEBM", "MP4", "WEBM", "MKV", "TS", "AAC", "H264", "H264_WS", "RAW_WS", "JSON_WS", "FLV", "HDS", "SMOOTHSTREAMING", "SDP", "MIST_HTML", "RTMP", "RTSP", "SRT", "DTSC":
		return mist.PlaybackProtocol(requested), nil
	default:
		return "", fmt.Errorf("unsupported viewer protocol")
	}
}

func viewerDemoForProtocol(response *sharedpb.ViewerEndpointResponse, protocol string) (*sharedpb.ViewerEndpointResponse, error) {
	if protocol == "" {
		return response, nil
	}
	var matches []*sharedpb.ViewerEndpoint
	for _, endpoint := range append([]*sharedpb.ViewerEndpoint{response.Primary}, response.Fallbacks...) {
		if endpoint == nil {
			continue
		}
		for key, output := range endpoint.Outputs {
			if output != nil && mist.PlaybackProtocol(key) == protocol {
				endpoint.Protocol, endpoint.Url = protocol, output.Url
				endpoint.Outputs = map[string]*sharedpb.OutputEndpoint{key: output}
				matches = append(matches, endpoint)
				break
			}
		}
	}
	if len(matches) == 0 {
		return nil, fmt.Errorf("requested viewer protocol is unavailable in demo")
	}
	response.Primary, response.Fallbacks = matches[0], matches[1:]
	return response, nil
}

File: npm_player/packages/core/src/core/ViewerProtocol.ts
import type { ContentEndpoints, EndpointInfo } from "../types";

export const viewerProtocols = {
  WEBRTC: "webrtc",
  WHEP: "whep",
  HLS: "hls",
  DASH: "dash",
  HLS_CMAF: "cmaf",
  MEWS: "wsmp4",
  MEWS_WEBM: "mews_webm",
  MP4: "mp4",
  WEBM: "webm",
  MKV: "mkv",
  TS: "ts",
  AAC: "aac",
  H264: "h264",
  H264_WS: "h264_ws",
  RAW_WS: "raw_ws",
  JSON_WS: "json_ws",
  FLV: "flv",
  HDS: "hds",
  SMOOTHSTREAMING: "smoothstreaming",
  SDP: "sdp",
  MIST_HTML: "mist_html",
  RTMP: "rtmp",
  RTSP: "rtsp",
  SRT: "srt",
  DTSC: "dtsc",
} as const;

export type ViewerProtocol = keyof typeof viewerProtocols;
const canonicalProtocols = new Set<string>(Object.values(viewerProtocols));

export function canonicalViewerProtocol(value: unknown): string | undefined {
  if (typeof value !== "string") return undefined;
  if (value.toUpperCase() === "MIST_WEBRTC") return "webrtc";
  return (
    viewerProtocols[value.toUpperCase() as ViewerProtocol] ??
    (canonicalProtocols.has(value) ? value : undefined)
  );
}

export function requireViewerProtocol(
  endpoints: ContentEndpoints,
  protocol: ViewerProtocol
): ContentEndpoints {
  const canonical = viewerProtocols[protocol];
  const primary = endpoints.primary;
  if (
    !canonical ||
    canonicalViewerProtocol(primary.protocol) !== canonical ||
    !primary.nodeId ||
    !validViewerURL(primary.url, canonical)
  ) {
    throw new Error("Gateway did not return the requested playback format");
  }
  const outputs: NonNullable<EndpointInfo["outputs"]> = {};
  for (const [key, output] of Object.entries(primary.outputs ?? {})) {
    const outputProtocol = canonicalViewerProtocol(key);
    const compatible =
      outputProtocol === canonical || (canonical === "mkv" && outputProtocol === "webm");
    if (compatible && output?.url === primary.url) outputs[key] = output;
  }
  if (!Object.keys(outputs).length) {
    throw new Error("Gateway did not return the requested playback output");
  }
  return { ...endpoints, primary: { ...primary, outputs }, fallbacks: [] };
}

function validViewerURL(value: string | undefined, protocol: string): boolean {
  if (!value) return false;
  try {
    const url = new URL(value);
    if (!url.hostname || url.username || url.password || url.hash) return false;
    if (["webrtc", "wsmp4", "mews_webm", "h264_ws", "raw_ws", "json_ws"].includes(protocol)) {
      return url.protocol === "ws:" || url.protocol === "wss:";
    }
    if (protocol === "rtmp") return url.protocol === "rtmp:" || url.protocol === "rtmps:";
    if (["rtsp", "srt", "dtsc"].includes(protocol)) return url.protocol === `${protocol}:`;
    return url.protocol === "http:" || url.protocol === "https:";
  } catch {
    return false;
  }
}
